### PR TITLE
[Merged by Bors] - Add macro to implement reflect for struct types and migrate glam types

### PR DESCRIPTION
**Pull request by PROMETHIA-27**
# Objective

Relevant issue: #4474

Currently glam types implement Reflect as a value, which is problematic for reflection, making scripting/editor work much more difficult. This PR re-implements them as structs.

## Solution

Added a new proc macro, `impl_reflect_struct`, which replaces `impl_reflect_value` and `impl_from_reflect_value` for glam types. This macro could also be used for other types, but I don't know of any that would require it. It's specifically useful for foreign types that cannot derive Reflect normally.

---

### Added
- `impl_reflect_struct` proc macro

### Changed
- Glam reflect impls have been replaced with `impl_reflect_struct`
- from_reflect's `impl_struct` altered to take an optional custom constructor, allowing non-default non-constructible foreign types to use it
- Calls to `impl_struct` altered to conform to new signature
- Altered glam types (All vec/mat combinations) have a different serialization structure, as they are reflected differently now.

## Migration Guide

This will break altered glam types serialized to RON scenes, as they will expect to be serialized/deserialized as structs rather than values now. A future PR to add custom serialization for non-value types is likely on the way to restore previous behavior. Additionally, calls to `impl_struct` must add a `None` parameter to the end of the call to restore previous behavior.

### CODE REVIEW

File: crates/bevy_reflect/src/impls/glam.rs
Comment (ItsDoot) [inline crates/bevy_reflect/src/impls/glam.rs:41]: If Quat is to be left as a value then this macro call and the one for DQuat should remain.

Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:41]: True, fixed

**Comment by MrGVSV** [comment]
I could be wrong, but I believe this would break current scene formats which expect a tuple-like value:

```rust
{
  "type": "glam::vec3::Vec3",
  "value": (1.0, 1.0, 1.0),
},
```

This would then become:

```rust
{
  "type": "glam::vec3::Vec3",
  "struct": {
    "x": {
      "type": "f32",
      "value": 1.0
    },
    // ...
  },
},
```

This is probably okay, but worth mentioning in the migration guide at least.

**Comment by PROMETHIA-27** [comment]
That makes sense. I'm having trouble confirming it from looking at bevy_serde, but I will try to test this soon. Will add it to the description.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:580]: Personally, I think this should be defined in a custom struct that impls `Parse`. It just helps make it clear what you're parsing and separates the structure of the macro from what it does.

Additionally, it would be a good place to document the macro's structure and parameters for internal usage.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:579]: I think we should add a doc comment here explaining to potential users what the macro looks like, how it's used, and what it does.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:581]: Another reason to use `Parse` is that if this fails, we don't get any help from the compiler (which might otherwise say something like `expected keyword: Constructor` or something.

Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:581]: Nit: `r#constructor` and `ctor` are a bit confusing together (`r#constructor` is already a bit confusing with the raw identifier). Maybe give these clearer names?

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:584]: I recently learned about the [`custom_keyword!`](https://docs.rs/syn/latest/syn/macro.custom_keyword.html) macro, which might be useful here and with `BevyReflectPath` (especially for letting the compiler know what should be expected).

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:585]: This (and the other panics) should probably return a compile error instead.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:617]: Nit: Same with `r#constructor` above and `r#struct` below, I think we can avoid using raw identifiers.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:628]: We should probably return a compile error explaining this to the user.

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/glam.rs:10]: I'm wondering if this would be better as a separate container attribute in a manner similar to #4140 where we can define something like `#[reflect(ctor = "Default::default()")]`.

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/glam.rs:11]: I'm not sure I understand what this is for. Wouldn't this be the same as using `BevyManifest`? Or is it doing something different?

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/glam.rs:16]: This is an interesting way of doing this. It makes sense, but it's not super clear (especially for potential users) that it's not creating the struct, just using it.

I think this would also make a good attribute candidate on its own: [`#[reflect(remote = "MockIVec2")]`](https://serde.rs/remote-derive.html).

**Review by MrGVSV — CHANGES_REQUESTED** [review]
Overall, I think this is a good change. I was torn at first because I like the one-line tuples defined in the scene format, but I realize now that we can just define custom serde implementations if we really wanted to keep them like that. So I'm on board!

Left a few suggestions and some points for discussion.

One last thing I think this PR really needs is a set of tests. Could you also add some of those?

File: crates/bevy_reflect/src/impls/glam.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:11]: I ran into a strange issue where specifically `use bevy_reflect::Struct` is ambiguous when called from a context where `use crate as bevy_reflect` is defined. This was not a problem before since we're only defining a Value reflect, but for a Struct reflect it needs to be either `self::bevy_reflect` or `crate::` to work, so I needed some way to override the path. I felt like this was a less intrusive option that messing with BevyManifest or infecting more macros than I already had to touch. If you have a better suggestion I would love to hear it! Not a big fan of either the Constructor() or BevyReflectPath() parts, but wasn't sure how best to get around it.

**Comment by PROMETHIA-27** [comment]
I've verified that this change alters how glam types are serialized like you mentioned. If they're serialized directly, they work as normal (e.g. using `serialize_ron()` from `bevy_scene::dynamic_scene`) but when using ReflectSerializer they are unfolded as complex struct definitions. The reflect serializer/deserializer likely only respects custom serializers when that type is a value, not when it's a struct. Not sure where to proceed with that from here; maybe letting struct reflect types check for a custom serialization scheme and use that if present? I feel like that deserves its own PR though.

**Comment by MrGVSV** [comment]
> The reflect serializer/deserializer likely only respects custom serializers when that type is a value, not when it's a struct. Not sure where to proceed with that from here; maybe letting struct reflect types check for a custom serialization scheme and use that if present? I feel like that deserves its own PR though.

Yep this is currently the case. I addressed this in #4561 to hopefully allow custom implementations to be used. But either way I think you’re okay not worrying about serialization 🙂

**Comment by PROMETHIA-27** [comment]
Nice! I'll keep working on the other suggestions in the meantime then.

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/glam.rs:11]: Hm, that's strange. I wonder why that's the case. So does it only affect the test modules (which is where I'm assuming `use crate as bevy_reflect` is used)?

And yeah, I think `#[reflect(path = "self::bevy_reflect", ctor = "Default::default()"]` is definitely an improvement. It's more consistent with the attributes we have now.

Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:11]: yep! I'm working on implementing those both once I get to them. I do think ctor could get a little weird if it gets complicated but that's probably still better than the weird tag thing I made.
And no, this affects the glam module where the glam types are impl'd for reflect. I feel like I may simply be not getting something but hopefully someone else figures that out before this is merged because I could not. Would be nice to just drop the path attribute entirely

File: crates/bevy_reflect/src/impls/glam.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:16]: I'm a bit conflicted on this. I do agree that it's not super clear but adding a remote and a Mock{Type} would expand it significantly and be a bit awkward to use. Maybe something more like this?
```
impl_reflect_struct_and_from_reflect_struct!(
    #[reflect(target = "IVec2", ctor/path, traits)
    struct MockIVec2 {
        x: i32,
        y: i32
    }
)
```
Do you think this is better?

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs]: How is that done? Make it emit an `error!` call?

Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs]: I like to create a [`syn::Error`](https://docs.rs/syn/latest/syn/parse/struct.Error.html) and then return `my_error.into_compile_error()`.

The great thing about `syn::Error` is that it allows you to [combine](https://docs.rs/syn/latest/syn/parse/struct.Error.html#method.combine) multiple errors into one, which allows the compiler to give the most amount of information back to the user.

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/glam.rs:16]: Yeah perhaps you're right. I never liked serde's remote approach either. I think your method with the target and stuff is probably more intuitive. Would it make sense to rename this macro to something like `mock_reflect_struct` since it's really only useful for that purpose?

One thing I don't like about this is the use of `target` in the `#[reflect]` since it might confuse users into thinking this is available outside this macro (which I don't think it is or will be). What do you think of a syntax like:

```rust
mock_reflect_struct! {
  #[mock = "IVec2"] // <- Required mock target
  #[reflect(ctor/path, traits)] // <- These should probably be optional
  struct {
    x: i32,
    y: i32
  } // <- Actual mock structure
}
```

Or something like that if you have any ideas. Or do you think this isn't the way to go here?

Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:16]: I think a rename is a good idea. The current name is descriptive but irritatingly long. However, this would also diverge further from `impl_reflect_value` and `impl_from_reflect_value`, which this is an alternative to and I would prefer to mirror as much as possible (unfortunately splitting the base impl and the from_reflect impl would require duplicating a lot of unnecessary code, so I merged them). I'm not certain that the separate mock attribute is better than a target attribute, I think it would just be more complicated. I do like removing the struct name, but that would also break the DeriveInput parser. That would be a real pain to put back together. Currently I believe the #[reflect] attribute is entirely optional, if you don't have anything to add to it.
And just to be clear, no, the mock struct is never actually constructed, it's just a description of what to pretend the target struct is defined as

Comment (MrGVSV) [reply crates/bevy_reflect/src/impls/glam.rs:16]: > I think a rename is a good idea. The current name is descriptive but irritatingly long. However, this would also diverge further from impl_reflect_value and impl_from_reflect_value, which this is an alternative to and I would prefer to mirror as much as possible (unfortunately splitting the base impl and the from_reflect impl would require duplicating a lot of unnecessary code, so I merged them). 

That makes sense. Yeah my reasoning for a rename (besides mocking being the main purpose it seems) was to also imply that we're not creating the struct, just using it. But maybe that's just not easily conveyed no matter the name (so we just need good documentation on the macro function to inform users of that). I’m still in favor of the mock naming, but if you have other ideas or pushback on that, we can bikeshed it a bit more.

> I do like removing the struct name, but that would also break the DeriveInput parser. That would be a real pain to put back together.

And yeah I realized after writing that that keeping a name on the struct would just be easier to parse so I think it's fine to keep it.

> I'm not certain that the separate mock attribute is better than a target attribute, I think it would just be more complicated.

You’re right that it just adds complexity. I think you've convinced me that the way you had it originally is probably the simplest:

```rust
mock_reflect_struct! { // Or whatever we end up calling it
  #[reflect(stuff)] // <- Still optional
  struct IVec2 { // Actual target name used
    x: i32,
    y: i32,
  }
}
```

Sorry for the loop around haha but I’m glad we discussed alternatives at the very least!

**Comment by PROMETHIA-27** [comment]
I think the PR is now in a good state to review again. I believe I've implemented all the previous suggestions. One lingering issue I have is that if you fail to provide a functioning ctor (e.g. if it defaults and the default is broken) it will throw something unintelligible as a result of failing to access fields, and not suggest anything related to the ctor tag. I'd like to have better errors for that but I have no idea how we could detect that the ctor code we're going to put in will error when compiled and override it. I don't believe it's possible but hopefully I'm missing something?

**Comment by PROMETHIA-27** [comment]
Oh and bikeshedding for a potential rename but I'm thinking that should be a separate PR that also applies to the impl_reflect_value/impl_from_reflect_value/etc. macros, for consistency.
Although, those aren't really "mocking" anything. 
Maybe just cut off the _and... portion and clarify in the docstring?

**Comment by MrGVSV** [comment]
> One lingering issue I have is that if you fail to provide a functioning ctor (e.g. if it defaults and the default is broken) it will throw something unintelligible as a result of failing to access fields, and not suggest anything related to the ctor tag.

Would #3733 help here? If so, it might be good to wait for that to land so we can get rid of `ctor` (and maybe add it back in through a separate PR).

**Comment by MrGVSV** [comment]
> Oh and bikeshedding for a potential rename but I'm thinking that should be a separate PR that also applies to the impl_reflect_value/impl_from_reflect_value/etc. macros, for consistency. Although, those aren't really "mocking" anything. Maybe just cut off the _and... portion and clarify in the docstring?

Yeah dropping the _and portion will probably look nicer. As for bikeshedding in a separate PR/issue, I think that should be okay. Ideally we would get that merged within the same release as this PR so we don't break any public uses of it.

**Comment by PROMETHIA-27** [comment]
#3733 would be enough for the purposes of migrating Glam types, and probably the vast majority of cases. I think it'd be worth it to wait for that. I'll cut off the _and portion now. I'm not too worried about when a rename PR hits, because this method is only useful internally to bevy because of orphan rules. For a user of bevy, a user-defined type can use a derive, and a foreign type can't use this macro. So any rename PR should be able to just use an IDE rename feature in bevy to hit all uses of the macro.

**Comment by MrGVSV** [comment]
> #3733 would be enough for the purposes of migrating Glam types, and probably the vast majority of cases. I think it'd be worth it to wait for that.

#3733 is merged! So you should be able to use that now (hopefully).

**Comment by PROMETHIA-27** [comment]
Ok, that's done now. I think we're ready for the next merge train? I think I'll investigate the path stuff again to see if there might be a better solution for that, would be nice to totally remove any special attributes.

**Comment by PROMETHIA-27** [comment]
Ah, it's a weirder error than I thought. It's actually doc tests failing. Without path overrides, cargo check passes and all normal tests pass but doc tests fail. Maybe there is a better solution.

**Comment by PROMETHIA-27** [comment]
All special attributes gone :) So much nicer

**Comment by MrGVSV** [comment]
> All special attributes gone :) So much nicer

Awesome! I'll check it out this weekend!

File: crates/bevy_reflect/src/impls/glam.rs
Comment (MrGVSV) [inline crates/bevy_reflect/src/impls/glam.rs:153]: We can come back to this in a future PR/issue if it's too contentious, but personally I think reflection should be allowed to break these read-only rules (just like it can for private fields).

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (MrGVSV) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:797]: Super-Nit: For consistency with the other methods, you could do:
```suggestion
    TokenStream::from(quote! {
        #impl_struct

        #impl_from_struct
    })
```

**Review by MrGVSV — COMMENTED** [review]
Looks good! I think the last few comments/suggestions are mainly just documentation-related.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:690]: Nope, replaced with an explanation about when #[reflect(Default)] is necessary.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:678]: Orphan rule is only an issue if you attempt to use this macro from outside bevy_reflect. In that case, you either need the type's owner to use #[derive(Reflect)] or you yourself need to use #[derive(Reflect)]. This macro is specifically for bevy to reflect standard foreign struct types, which so far is exclusively the glam types.

Though, this gives me an idea; wouldn't this macro work if applied to a newtype wrapper that has #[derive(Deref)], and you inform it of the layout of the inner, presumably foreign type? I'll have to try that. Could be a useful if niche trick

File: crates/bevy_reflect/src/impls/glam.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/impls/glam.rs:153]: I don't even understand how glam made those fields read-only, to be honest. If we want to break those rules I think we need to fork/PR glam itself. We could express those rules without breaking them, to allow read-only reflection, without doing so but I'm leaving that to a separate PR.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (PROMETHIA-27) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:678]: Ha, works perfectly. That's some emergent design
```
#[derive(prelude::Deref, prelude::DerefMut, Default)]
struct BevyDMat4(self::math::DMat4);

use reflect::std_traits::ReflectDefault;
use prelude::Reflect;
use math::DVec4;

reflect::impl_reflect_struct!(
    #[reflect(Default)]
    struct BevyDMat4 {
        x_axis: DVec4,
        y_axis: DVec4,
        z_axis: DVec4,
        w_axis: DVec4
    }
);

#[test]
fn test() {
    let mut b = BevyDMat4(Default::default());
    b.x_axis.x = 12.0;

    let refl: &dyn Reflect = &b;

    use reflect::GetPath;
    let val = refl.path("x_axis.x").unwrap();
    assert_eq!(val.downcast_ref::<f64>().unwrap(), &12.0);
}
```

Comment (MrGVSV) [reply crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:678]: Ooh interesting. This might be fairly useful actually since foreign types are a bit of an issue. It may even be worth adding an example for in `examples/reflection` (in a separate PR, of course).

**Review by MrGVSV — APPROVED** [review]

File: crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:11]:
```suggestion
    custom_constructor: Option<proc_macro2::TokenStream>,
```

File: crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:64]:
```suggestion
    let constructor = if let Some(ctor) = custom_constructor {
```

File: crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:92]:
```suggestion
                    #constructor
```

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
This looks good. One nit about naming conventions; I didn't recognize `ctor` at a glance.

File: crates/bevy_reflect/bevy_reflect_derive/src/lib.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:766]:
```suggestion
    let constructor = if attrs
```

File: crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:64]:
```suggestion
    let constructor = if let Some(constructor) = custom_constructor {
```

File: crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs
Comment (alice-i-cecile) [inline crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:66]:
```suggestion
            let mut value: Self = #constructor;
```

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**
